sql/core/src/main/scala/org/apache/spark/sql/execution/joins/SortMergeOuterJoin.scala

```diff
@@ -271,3 +300,243 @@ private class RightOuterIterator(
 
   override def getRow: InternalRow = resultProj(joinedRow)
 }
+
+private object SortMergeFullJoinScanner {
+  object Status extends Enumeration {
+    type Status = Value
+    val Init, LeftAdvanced, RightAdvanced, NoAdvanced, Buffering = Value
```

Comment:
It's hard to manage these statuses correctly, I think we could get rid of LeftAdvanced and RightAdvanced by using EmptyRow as Matched rows. Also we could have proper initialization to get rid of Init. Then only `NoAdvanced` and `Buffering` left, it means we only need `false` and `true`.